src/trs/hmtrs.rs

```diff
@@ -246,71 +285,45 @@ impl HMTRS {
             .into_iter()
             .filter(|(o, _)| o == &Some(term.head()))
             .collect();
-        // fail if not exactly 1 match
-        if matches.len() == 0 {
-            return Ok(NEG_INFINITY);
-        }
         // process the match
-        match matches[0] {
-            (Some(Atom::Variable(_)), _) => Ok(lp),
-            (Some(Atom::Operator(_)), ref bts) => {
+        match matches.get(0) {
+            Some((Some(Atom::Variable(_)), _)) => Ok(lp),
+            Some((Some(Atom::Operator(_)), ref bts)) => {
                 for (subterm, body_type) in term.args().iter().zip(bts) {
                     let subtype = body_type.apply(ctx);
                     let subschema = TypeSchema::Monotype(subtype.clone());
-                    let tmp_lp = self.lp_term(subterm, &subschema, ctx, invent)?;
+                    let tmp_lp = self.logprior_term(subterm, &subschema, ctx, invent)?;
                     lp += tmp_lp;
-                    let final_schema = self.infer_term(subterm, ctx)
-                        .or(Err(SampleError("untypable".to_string())))?;
+                    let final_schema = self.infer_term(subterm, ctx)?;
                     let final_type = final_schema.instantiate(ctx);
                     if ctx.unify(&subtype, &final_type).is_err() {
                         return Ok(NEG_INFINITY);
                     }
                 }
-                return Ok(lp);
+                Ok(lp)
             }
-            _ => Err(SampleError("Should never happen -- FIXME!".to_string())),
+            Some(_) => panic!("Should never happen -- FIXME!"),
```

Comment:
This `panic` is another awkward result of the way I wrote this initially. Any ideas for cleanly removing it?

Reply:
Yeah, I intend on taking a closer look to making that state unrepresentable.